File: buildx-web/src/util/planarRange.ts
import type { Text } from "@codemirror/state";

export type PlanarRange = {
  fromRow: number;
  fromColumn: number;
  toRow: number;
  toColumn: number;
  tabWidth?: number;
};

export const SOURCE_POSITION_PREFIX = "source-";

export function planarRangeToString(range: PlanarRange): string {
  const tabWidth = range.tabWidth && range.tabWidth > 0 ? range.tabWidth : 1;
  return `${range.fromRow}.${range.fromColumn}-${range.toRow}.${range.toColumn}-${tabWidth}`;
}

export function sourcePositionFromRange(range: PlanarRange): string {
  return `${SOURCE_POSITION_PREFIX}${planarRangeToString(range)}`;
}

export function parsePlanarRange(value: string): PlanarRange | null {
  const parts = value.split("-");
  if (parts.length < 2 || parts.length > 3) {
    return null;
  }
  const [fromRow, fromColumn] = parts[0].split(".").map(Number);
  const [toRow, toColumn] = parts[1].split(".").map(Number);
  if ([fromRow, fromColumn, toRow, toColumn].some((n) => Number.isNaN(n))) {
    return null;
  }
  const tabWidth = parts.length === 3 ? Number(parts[2]) : 1;
  return { fromRow, fromColumn, toRow, toColumn, tabWidth: Number.isNaN(tabWidth) ? 1 : tabWidth };
}

export function parseSourcePosition(position: string | null | undefined): PlanarRange | null {
  if (!position || !position.startsWith(SOURCE_POSITION_PREFIX)) {
    return null;
  }
  return parsePlanarRange(position.slice(SOURCE_POSITION_PREFIX.length));
}

/** Normalize a CodeMirror selection into OneDev's planar range (0-based lines/chars). */
export function rangeFromSelection(doc: Text, anchor: number, head: number): PlanarRange | null {
  const from = Math.min(anchor, head);
  const to = Math.max(anchor, head);
  if (from === to) {
    return null;
  }

  const fromLine = doc.lineAt(from);
  const toLine = doc.lineAt(to);

  let fromRow = fromLine.number - 1;
  let fromColumn = from - fromLine.from;
  let toRow = toLine.number - 1;
  let toColumn = to - toLine.from;

  if (toColumn === 0 && toRow > fromRow) {
    const prevLine = doc.line(toLine.number - 1);
    toRow = prevLine.number - 1;
    toColumn = prevLine.length;
  }

  return { fromRow, fromColumn, toRow, toColumn, tabWidth: 1 };
}

export function rangeToPositions(
  doc: Text,
  range: PlanarRange,
): { from: number; to: number } {
  const fromLine = doc.line(range.fromRow + 1);
  const toLine = doc.line(range.toRow + 1);
  const from = fromLine.from + range.fromColumn;
  let to = toLine.from + range.toColumn;
  if (to < from) {
    to = from;
  }
  return { from, to };
}

File: buildx-server/internal/model/planar_range.go
package model

import (
	"fmt"
	"strconv"
	"strings"
)

const SourcePositionPrefix = "source-"

// PlanarRange is a zero-based line/column selection in a text file.
// String format matches OneDev: "{fromRow}.{fromColumn}-{toRow}.{toColumn}-{tabWidth}".
type PlanarRange struct {
	FromRow    int `json:"fromRow"`
	FromColumn int `json:"fromColumn"`
	ToRow      int `json:"toRow"`
	ToColumn   int `json:"toColumn"`
	TabWidth   int `json:"tabWidth,omitempty"`
}

func (r PlanarRange) String() string {
	tabWidth := r.TabWidth
	if tabWidth <= 0 {
		tabWidth = 1
	}
	return fmt.Sprintf("%d.%d-%d.%d-%d", r.FromRow, r.FromColumn, r.ToRow, r.ToColumn, tabWidth)
}

// SourcePosition returns the blob page position query value.
func (r PlanarRange) SourcePosition() string {
	return SourcePositionPrefix + r.String()
}

// ParseSourcePosition parses a blob position query value into a range.
func ParseSourcePosition(position string) (*PlanarRange, error) {
	if position == "" {
		return nil, nil
	}
	if !strings.HasPrefix(position, SourcePositionPrefix) {
		return nil, fmt.Errorf("invalid source position prefix")
	}
	return ParsePlanarRange(position[len(SourcePositionPrefix):])
}

// ParsePlanarRange parses OneDev's planar range string.
func ParsePlanarRange(s string) (*PlanarRange, error) {
	parts := strings.Split(s, "-")
	if len(parts) < 2 || len(parts) > 3 {
		return nil, fmt.Errorf("invalid planar range: %q", s)
	}

	fromParts := strings.Split(parts[0], ".")
	if len(fromParts) != 2 {
		return nil, fmt.Errorf("invalid planar range from: %q", parts[0])
	}
	toParts := strings.Split(parts[1], ".")
	if len(toParts) != 2 {
		return nil, fmt.Errorf("invalid planar range to: %q", parts[1])
	}

	fromRow, err := strconv.Atoi(fromParts[0])
	if err != nil {
		return nil, fmt.Errorf("invalid from row: %w", err)
	}
	fromCol, err := strconv.Atoi(fromParts[1])
	if err != nil {
		return nil, fmt.Errorf("invalid from column: %w", err)
	}
	toRow, err := strconv.Atoi(toParts[0])
	if err != nil {
		return nil, fmt.Errorf("invalid to row: %w", err)
	}
	toCol, err := strconv.Atoi(toParts[1])
	if err != nil {
		return nil, fmt.Errorf("invalid to column: %w", err)
	}

	tabWidth := 1
	if len(parts) == 3 {
		tabWidth, err = strconv.Atoi(parts[2])
		if err != nil {
			return nil, fmt.Errorf("invalid tab width: %w", err)
		}
	}

	return &PlanarRange{
		FromRow:    fromRow,
		FromColumn: fromCol,
		ToRow:      toRow,
		ToColumn:   toCol,
		TabWidth:   tabWidth,
	}, nil
}
